Make only one function that needs to be implemented when searching for types

clayborg opened the pull request:
This patch revives the effort to get this Phabricator patch into upstream:

https://reviews.llvm.org/D137900

This patch was accepted before in Phabricator but I found some -gsimple-template-names issues that are fixed in this patch.

A fixed up version of the description from the original patch starts now.

This patch started off trying to fix Module::FindFirstType() as it sometimes didn't work. The issue was the SymbolFile plug-ins didn't do any filtering of the matching types they produced, and they only looked up types using the type basename. This means if you have two types with the same basename, your type lookup can fail when only looking up a single type. We would ask the Module::FindFirstType to lookup "Foo::Bar" and it would ask the symbol file to find only 1 type matching the basename "Bar", and then we would filter out any matches that didn't match "Foo::Bar". So if the SymbolFile found "Foo::Bar" first, then it would work, but if it found "Baz::Bar" first, it would return only that type and it would be filtered out.

Discovering this issue lead me to think of the patch Alex Langford did a few months ago that was done for finding functions, where he allowed SymbolFile objects to make sure something fully matched before parsing the debug information into an AST type and other LLDB types. So this patch aimed to allow type lookups to also be much more efficient.

As LLDB has been developed over the years, we added more ways to to type lookups. These functions have lots of arguments. This patch aims to make one API that needs to be implemented that serves all previous lookups:

- Find a single type
- Find all types
- Find types in a namespace

This patch introduces a `TypeQuery` class that contains all of the state needed to perform the lookup which is powerful enough to perform all of the type searches that used to be in our API. It contain a vector of CompilerContext objects that can fully or partially specify the lookup that needs to take place.

If you just want to lookup all types with a matching basename, regardless of the containing context, you can specify just a single CompilerContext entry that has a name and a CompilerContextKind mask of CompilerContextKind::AnyType.

Or you can fully specify the exact context to use when doing lookups like: CompilerContextKind::Namespace "std"
CompilerContextKind::Class "foo"
CompilerContextKind::Typedef "size_type"

This change expands on the clang modules code that already used a vector<CompilerContext> items, but it modifies it to work with expression type lookups which have contexts, or user lookups where users query for types. The clang modules type lookup is still an option that can be enabled on the `TypeQuery` objects.

This mirrors the most recent addition of type lookups that took a vector<CompilerContext> that allowed lookups to happen for the expression parser in certain places.

Prior to this we had the following APIs in Module:

```
void
Module::FindTypes(ConstString type_name, bool exact_match, size_t max_matches,
                  llvm::DenseSet<lldb_private::SymbolFile *> &searched_symbol_files,
                  TypeList &types);

void
Module::FindTypes(llvm::ArrayRef<CompilerContext> pattern, LanguageSet languages,
                  llvm::DenseSet<lldb_private::SymbolFile *> &searched_symbol_files,
                  TypeMap &types);

void Module::FindTypesInNamespace(ConstString type_name,
                                  const CompilerDeclContext &parent_decl_ctx,
                                  size_t max_matches, TypeList &type_list);
```

The new Module API is much simpler. It gets rid of all three above functions and replaces them with:

```
void FindTypes(const TypeQuery &query, TypeResults &results);
```
The `TypeQuery` class contains all of the needed settings:

- The vector<CompilerContext> that allow efficient lookups in the symbol file classes since they can look at basename matches only realize fully matching types. Before this any basename that matched was fully realized only to be removed later by code outside of the SymbolFile layer which could cause many types to be realized when they didn't need to.
- If the lookup is exact or not. If not exact, then the compiler context must match the bottom most items that match the compiler context, otherwise it must match exactly
- If the compiler context match is for clang modules or not. Clang modules matches include a Module compiler context kind that allows types to be matched only from certain modules and these matches are not needed when d oing user type lookups.
- An optional list of languages to use to limit the search to only certain languages

The `TypeResults` object contains all state required to do the lookup and store the results:
- The max number of matches
- The set of SymbolFile objects that have already been searched
- The matching type list for any matches that are found

The benefits of this approach are:
- Simpler API, and only one API to implement in SymbolFile classes
- Replaces the FindTypesInNamespace that used a CompilerDeclContext as a way to limit the search, but this only worked if the TypeSystem matched the current symbol file's type system, so you couldn't use it to lookup a type in another module
- Fixes a serious bug in our FindFirstType functions where if we were searching for "foo::bar", and we found a "baz::bar" first, the basename would match and we would only fetch 1 type using the basename, only to drop it from the matching list and returning no results

clayborg committed:
Make only one function that needs to be implemented when searching for types.

This patch revives the effort to get this Phabricator patch into upstream:

https://reviews.llvm.org/D137900

This patch was accepted before in Phabricator but I found some -gsimple-template-names issues that are fixed in this patch.

A fixed up version of the description from the original patch starts now.

This patch started off trying to fix Module::FindFirstType() as it sometimes didn't work. The issue was the SymbolFile plug-ins didn't do any filtering of the matching types they produced, and they only looked up types using the type basename. This means if you have two types with the same basename, your type lookup can fail when only looking up a single type. We would ask the Module::FindFirstType to lookup "Foo::Bar" and it would ask the symbol file to find only 1 type matching the basename "Bar", and then we would filter out any matches that didn't match "Foo::Bar". So if the SymbolFile found "Foo::Bar" first, then it would work, but if it found "Baz::Bar" first, it would return only that type and it would be filtered out.

Discovering this issue lead me to think of the patch Alex Langford did a few months ago that was done for finding functions, where he allowed SymbolFile objects to make sure something fully matched before parsing the debug information into an AST type and other LLDB types. So this patch aimed to allow type lookups to also be much more efficient.

As LLDB has been developed over the years, we added more ways to to type lookups. These functions have lots of arguments. This patch aims to make one API that needs to be implemented that serves all previous lookups:

- Find a single type
- Find all types
- Find types in a namespace

This patch introduces a `TypeQuery` class that contains all of the state needed to perform the lookup which is powerful enough to perform all of the type searches that used to be in our API. It contain a vector of CompilerContext objects that can fully or partially specify the lookup that needs to take place.

If you just want to lookup all types with a matching basename, regardless of the containing context, you can specify just a single CompilerContext entry that has a name and a CompilerContextKind mask of CompilerContextKind::AnyType.

Or you can fully specify the exact context to use when doing lookups like:
CompilerContextKind::Namespace "std"
CompilerContextKind::Class "foo"
CompilerContextKind::Typedef "size_type"

This change expands on the clang modules code that already used a vector<CompilerContext> items, but it modifies it to work with expression type lookups which have contexts, or user lookups where users query for types. The clang modules type lookup is still an option that can be enabled on the `TypeQuery` objects.

This mirrors the most recent addition of type lookups that took a vector<CompilerContext> that allowed lookups to happen for the expression parser in certain places.

Prior to this we had the following APIs in Module:

```
void
Module::FindTypes(ConstString type_name, bool exact_match, size_t max_matches,
                  llvm::DenseSet<lldb_private::SymbolFile *> &searched_symbol_files,
                  TypeList &types);

void
Module::FindTypes(llvm::ArrayRef<CompilerContext> pattern, LanguageSet languages,
                  llvm::DenseSet<lldb_private::SymbolFile *> &searched_symbol_files,
                  TypeMap &types);

void Module::FindTypesInNamespace(ConstString type_name,
                                  const CompilerDeclContext &parent_decl_ctx,
                                  size_t max_matches, TypeList &type_list);
```

The new Module API is much simpler. It gets rid of all three above functions and replaces them with:

```
void FindTypes(const TypeQuery &query, TypeResults &results);
```
The `TypeQuery` class contains all of the needed settings:

- The vector<CompilerContext> that allow efficient lookups in the symbol file classes since they can look at basename matches only realize fully matching types. Before this any basename that matched was fully realized only to be removed later by code outside of the SymbolFile layer which could cause many types to be realized when they didn't need to.
- If the lookup is exact or not. If not exact, then the compiler context must match the bottom most items that match the compiler context, otherwise it must match exactly
- If the compiler context match is for clang modules or not. Clang modules matches include a Module compiler context kind that allows types to be matched only from certain modules and these matches are not needed when d oing user type lookups.
- An optional list of languages to use to limit the search to only certain languages

The `TypeResults` object contains all state required to do the lookup and store the results:
- The max number of matches
- The set of SymbolFile objects that have already been searched
- The matching type list for any matches that are found

The benefits of this approach are:
- Simpler API, and only one API to implement in SymbolFile classes
- Replaces the FindTypesInNamespace that used a CompilerDeclContext as a way to limit the search, but this only worked if the TypeSystem matched the current symbol file's type system, so you couldn't use it to lookup a type in another module
- Fixes a serious bug in our FindFirstType functions where if we were searching for "foo::bar", and we found a "baz::bar" first, the basename would match and we would only fetch 1 type using the basename, only to drop it from the matching list and returning no results

clayborg commented:
@adrian-prantl This patch is exactly the same as it was back when you accepted it in Phabricator with two things added:
- fixes for -gsimple-template-names
- updated the new `SBType::FindDirectNestedType(...)`

All tests pass on macOS.

felipepiovezan commented:
I am really struggling to handle the scope of the changes here, IIUC there are a handful of changes that could be split into separate commits and merged independently of each other: there is some code being moved around, new GetCompilerContext APIs, the new query, the replacing of the old queries with the new queries. All of these could be different changes? It would also making reverting things a lot easier if we end up breaking bots down the line, and some of these changes are just "NFC goodness" that would never need to be reverted

adrian-prantl commented:
Sorry, it's been a while since I last reviewed this and I must admit I forgot about this in the mean time. The only thing I'd like to check before we can land this is to hear from @Michael137 who has been actively working on FindTypes performance recently whether this aligns or clashes with what he's been working on so we don't get a mid-air collision between patches.

clayborg commented:
> Sorry, it's been a while since I last reviewed this and I must admit I forgot about this in the mean time. The only thing I'd like to check before we can land this is to hear from @Michael137 who has been actively working on FindTypes performance recently whether this aligns or clashes with what he's been working on so we don't get a mid-air collision between patches.

Sounds good, I look forward to hearing if this fixes the issues as well. Before this fix I ran a case where I did an expression like "MyClass::iterator" and before this fix, with linux and no accelerator tables, we would convert DWARF into clang AST types for over 2 million types, and after this fix it went down to 11.

hawkinsw commented on lldb/include/lldb/Symbol/CompilerDeclContext.h:
```diff
@@ -56,6 +57,13 @@ class CompilerDeclContext {
+  ///
+  /// \returns A valid vector of CompilerContext entries that describes
+  /// this declaration context. The first entry in the vector is the parent of
+  /// the subsequent entry, so the top most entry is the global namespace.
```
```suggestion
  /// the subsequent entry, so the topmost entry is the global namespace.
```

hawkinsw commented on lldb/include/lldb/Symbol/CompilerDecl.h:
```diff
@@ -84,6 +84,13 @@ class CompilerDecl {
+  ///
+  /// \returns A valid vector of CompilerContext entries that describes
+  /// this declaration. The first entry in the vector is the parent of
+  /// the subsequent entry, so the top most entry is the global namespace.
```
```suggestion
  /// the subsequent entry, so the topmost entry is the global namespace.
```

hawkinsw commented on lldb/include/lldb/Core/ModuleList.h:
```diff
@@ -340,26 +340,22 @@ class ModuleList {
-  /// \param[out] types
-  ///     A type list gets populated with any matches.
+  /// \param[in] query
+  ///     A type matching object that contains all of the details of the type
```
```suggestion
  ///     A type-matching object that contains all of the details of the type
```

hawkinsw commented on lldb/include/lldb/Core/Module.h:
```diff
@@ -415,70 +415,19 @@ class Module : public std::enable_shared_from_this<Module>,
-  ///     The name of a type within a namespace that should not include
-  ///     any qualifying namespaces (just a type basename).
+  /// \param[in] query
+  ///     A type matching object that contains all of the details of the type
```
```suggestion
  ///     A type-matching object that contains all of the details of the type
```

hawkinsw commented on lldb/include/lldb/Symbol/SymbolFile.h:
```diff
@@ -301,21 +301,20 @@ class SymbolFile : public PluginInterface {
-            llvm::DenseSet<lldb_private::SymbolFile *> &searched_symbol_files,
-            TypeMap &types);
+
+  /// Find types using a type matching object that contains all search
```
```suggestion
  /// Find types using a type-matching object that contains all search
```

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
 
+FLAGS_ENUM(TypeQueryOptions){
+    e_none = 0u,
+    /// If set TypeQuery::m_context contains an exact context that must match
```
```suggestion
    /// If set, TypeQuery::m_context contains an exact context that must match
```

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+FLAGS_ENUM(TypeQueryOptions){
+    e_none = 0u,
+    /// If set TypeQuery::m_context contains an exact context that must match
+    /// the full context. If not set TypeQuery::m_context can contain a partial
```
```suggestion
    /// the full context. If not set, TypeQuery::m_context can contain a partial
```

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+/// Using a TypeQuery class for matching types simplifies the internal APIs we
+/// need to implement type lookups in LLDB. Type lookups can fully specify the
+/// exact typename by filling out a complete or partial CompilerContext array.
+/// This allows for powerful searches and also allows the SymbolFile classes to
```
```suggestion
/// This technique allows for powerful searches and also allows the SymbolFile classes to
```

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+public:
+  TypeQuery() = delete;
+
+  /// Construct a type match object using a fully or partially qualified name.
```
```suggestion
  /// Construct a type-match object using a fully- or partially-qualified name.
```

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+  ///     named "foo".
+  ///
+  /// \param[in] type_name
+  ///   A fully or partially qualified type name. This name will be parsed and
```
```suggestion
  ///   A fully- or partially-qualified type name. This name will be parsed and
```

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+  ///   the name. This typename can be prefixed with "struct ", "class ",
+  ///   "union", "enum " or "typedef " before the actual type name to limit the
+  ///   results of the types that match. The declaration context can be
+  ///   specified with the "::" string. like "a::b::my_type".
```
```suggestion
  ///   specified with the "::" string. For example, "a::b::my_type".
```

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+  ///   TypeQueryOptions enumerations. \see TypeQueryOptions.
+  TypeQuery(llvm::StringRef name, TypeQueryOptions options = e_none);
+
+  /// Construct a type match object that matches a type basename that exists
```
```suggestion
  /// Construct a type-match object that matches a type basename that exists
```

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+  ///
+  /// \param[in] decl_ctx
+  ///   A declaration context object that comes from a TypeSystem plug-in. This
+  ///   object will be asked to full the array of CompilerContext objects
```
Seems like we are missing a word here.

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+  ///   TypeQueryOptions enumerations. \see TypeQueryOptions.
+  TypeQuery(const CompilerDeclContext &decl_ctx, ConstString type_basename,
+            TypeQueryOptions options = e_none);
+  /// Construct a type match object using a compiler declaration that specifies
```
```suggestion
  /// Construct a type-match object using a compiler declaration that specifies
```

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+  ///   TypeQueryOptions enumerations. \see TypeQueryOptions.
+  TypeQuery(const CompilerDecl &decl, TypeQueryOptions options = e_none);
+
+  /// Construct a type match object using a CompilerContext array.
```
```suggestion
  /// Construct a type-match object using a CompilerContext array.
```

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+  /// contexts specify contexts more completely to find exact matches within
+  /// clang module debug information. They will include the modules that the
+  /// type is included in and any functions that the type might be defined in.
+  /// This allows very fine grained type resolution.
```
```suggestion
  /// This allows very fine-grained type resolution.
```

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+  TypeQuery(const llvm::ArrayRef<lldb_private::CompilerContext> &context,
+            TypeQueryOptions options = e_none);
+
+  /// Construct a type match object that duplicates all matching criterea,
```
```suggestion
  /// Construct a type-match object that duplicates all matching criteria,
```

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+  /// but not any searched symbol files or the type map for matches. This allows
+  /// the m_context to be modified prior to performing another search.
+  TypeQuery(const TypeQuery &rhs) = default;
+  /// Assign a type match object that duplicates all matching criterea,
```
```suggestion
  /// Assign a type-match object that duplicates all matching criteria,
```

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+  ///   True if the context matches, false if it doesn't. If e_exact_match
+  ///   is set in m_options, then \a context must exactly match \a m_context. If
+  ///   e_exact_match is not set, then the bottom m_context.size() objects in
+  ///   the \a context must match. This allows SymbolFile objects the fill in a
```
*the* does not seem right here, but I am not sure what to put instead ... sorry!

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+  /// The \a m_context can be used in two ways: normal types searching with
+  /// the context containing a stanadard declaration context for a type, or
+  /// with the context being more complete for exact matches in clang modules.
+  /// Se this to true if you wish to search for a type in clang module.
```
```suggestion
  /// Set this to true if you wish to search for a type in clang module.
```

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+protected:
+  /// A full or partial compiler context array where the parent declaration
+  /// contexts appear at the top of the array starting at index zero and the
+  /// last entry is contains the type and name of the type we are looking for.
```
```suggestion
  /// last entry contains the type and name of the type we are looking for.
```

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+  /// An options bitmask that contains enabled options for the type query.
+  /// \see TypeQueryOptions.
+  TypeQueryOptions m_options;
+  /// If this has a value, then the language family must match at least one of
```
```suggestion
  /// If this variable has a value, then the language family must match at least one of
```

hawkinsw commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+  /// \see TypeQueryOptions.
+  TypeQueryOptions m_options;
+  /// If this has a value, then the language family must match at least one of
+  /// the specified languages. If this has no value, then the language of the
```
```suggestion
  /// the specified languages. If this variable has no value, then the language of the
```

hawkinsw reviewed (commented):
Thank you for the great work. I really appreciate the excellent documentation that you wrote. I am not well versed enough in the code to comment on the implementation, but your comments were so good that I intuitively understood the point of the new classes. I hope that some of these suggestions are helpful. Again, these are suggestions and you should not feel like you are under any obligation to accept them. 

Thanks for the great work!

Michael137 commented:
> > Sorry, it's been a while since I last reviewed this and I must admit I forgot about this in the mean time. The only thing I'd like to check before we can land this is to hear from @Michael137 who has been actively working on FindTypes performance recently whether this aligns or clashes with what he's been working on so we don't get a mid-air collision between patches.
> 
> Sounds good, I look forward to hearing if this fixes the issues as well. Before this fix I ran a case where I did an expression like "MyClass::iterator" and before this fix, with linux and no accelerator tables, we would convert DWARF into clang AST types for over 2 million types, and after this fix it went down to 11. This was when debugging LLDB + LLVM + clang with debug info, so there were TONS of "iterator" entries in the manual DWARF index.
> 
> On Apple the type accelerator tables have the hash of the fully qualified name which can help reduce the number of types we parse, but we don't have this on linux even with DWARF5 + .debug_names

To add more context to Adrian's comment, we've been working on reviving https://reviews.llvm.org/D101950, which changes the way we complete record types. The main blocker was performance, because as you mention, we often end up doing these expensive namespace searches, which gets much more noticeable with D101950. The problem we saw was mostly triggered by the second lookup that `-gsimple-template-names` does in `FindTypes`.

I applied your patch on top of D101950 and performance seems to be much better (by 2 orders of magnitude). It did break some tests locally for me, so I'm just skimming this patch to see what changed

Michael137 commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+  ///
+  /// \param[in] options A set of boolean enumeration flags from the
+  ///   TypeQueryOptions enumerations. \see TypeQueryOptions.
+  TypeQuery(llvm::StringRef name, TypeQueryOptions options = e_none);
```
The comments refer to a parameter called `type_name`, but it's called `name`

Michael137 commented on lldb/source/API/SBModule.cpp:
```diff
@@ -437,26 +437,25 @@ lldb::SBType SBModule::FindFirstType(const char *name_cstr) {
   LLDB_INSTRUMENT_VA(this, name_cstr);
 
   ModuleSP module_sp(GetSP());
-  if (!name_cstr || !module_sp)
```
why did we get rid of the early return here?

Michael137 commented on lldb/source/API/SBModule.cpp:
```diff
@@ -437,26 +437,25 @@ lldb::SBType SBModule::FindFirstType(const char *name_cstr) {
+    module_sp->FindTypes(query, results);
+    TypeSP type_sp = results.GetFirstType();
+    if (type_sp)
+      return SBType(type_sp);
```
We previously would only return `sb_type` if `sb_type.IsValid() == true`. Does this mean we can now return an "invalid" SBType? Or is this checked elsewhere now?

Michael137 commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -45,6 +66,286 @@ struct CompilerContext {
+  ///     matches "foo".
+  ///   "class foo::bar": Only match types that are classes whose basename
+  ///     matches "bar" and that are contained in any named declaration context
+  ///     named "foo".
```
Can we also document the special behaviour of a query that begins with `::`? It looks like in those cases we set `exact_match` to `true`?

Michael137 commented on lldb/source/API/SBModule.cpp:
```diff
@@ -502,11 +499,8 @@ lldb::SBTypeList SBModule::FindTypes(const char *type) {
-          retval.Append(SBType(type_sp));
-      }
+      for (const TypeSP &type_sp : results.GetTypeMap().Types())
+        retval.Append(SBType(type_sp));
```
Should we add a debug-assert here (or maybe in `InsertUnique`) that tests `type_sp != nullptr`?

Michael137 commented on lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp:
```diff
@@ -9070,6 +9070,66 @@ size_t TypeSystemClang::DeclGetFunctionNumArguments(void *opaque_decl) {
 }
 
+static CompilerContextKind GetCompilerKind(clang::Decl::Kind clang_kind,
+                                           clang::DeclContext *decl_ctx) {
```
```suggestion
                                           clang::DeclContext const *decl_ctx) {
```

Michael137 commented on lldb/test/API/lang/cpp/unique-types4/TestUniqueTypes4.py:
```diff
@@ -12,33 +12,32 @@ class UniqueTypesTestCase4(TestBase):
         """Test that we display the correct template instantiation."""
         self.build(dictionary=debug_flags)
+        exe = self.getBuildArtifact("a.out")
+        print(exe)
```
leftover from debugging?

clayborg replied:
yes, I will remove

Michael137 commented on lldb/test/API/lang/cpp/unique-types4/TestUniqueTypes4.py:
```diff
@@ -12,33 +12,32 @@ class UniqueTypesTestCase4(TestBase):
         # FIXME: these should successfully print the values
         self.expect(
-            "expression ns::Foo<double>::value", substrs=["no member named"], error=True
+            "print ns::Foo<double>::value", substrs=["no template named 'Foo' in namespace 'ns'"], error=True
```
why change these to `print`?

clayborg replied:
switched them back to use `expression`

Michael137 commented on lldb/test/API/lang/cpp/unique-types4/TestUniqueTypes4.py:
```diff
@@ -12,33 +12,32 @@ class UniqueTypesTestCase4(TestBase):
-            "expression ns::FooDouble::value",
-            substrs=["Couldn't look up symbols"],
-            error=True,
+            "print ns::FooDouble::value", substrs=["(double) 0"]
```
this can be now turned into `expect_expr`

Michael137 commented on lldb/test/API/lang/cpp/unique-types4/TestUniqueTypes4.py:
```diff
@@ -12,33 +12,32 @@ class UniqueTypesTestCase4(TestBase):
-            "expression ns::FooInt::value",
-            substrs=["Couldn't look up symbols"],
-            error=True,
+            "print ns::FooInt::value", substrs=["(int) 0"]
```
`expect_expr`

Michael137 commented on lldb/source/Symbol/Type.cpp:
```diff
@@ -64,6 +64,125 @@ bool lldb_private::contextMatches(llvm::ArrayRef<CompilerContext> context_chain,
+}
+
+bool TypeResults::AlreadySearched(lldb_private::SymbolFile *sym_file) {
+  return !m_searched_symbol_files.insert(sym_file).second;
```
```suggestion
  return m_searched_symbol_files.contains(sym_file);
```

clayborg replied:
We actually want to try and insert it as this function is called as something that puts the symbol file into the map and returns `false` if it wasn't in the set. Or it tries to add it and returns `true` if it was alraedy in there. I have modified the header documentation to make this more clear.

Michael137 replied:
Ah makes sense

Michael137 commented on lldb/source/Plugins/ExpressionParser/Clang/ClangASTSource.cpp:
```diff
@@ -263,13 +252,6 @@ TagDecl *ClangASTSource::FindCompleteType(const TagDecl *decl) {
-      // We have found a type by basename and we need to make sure the decl
-      // contexts are the same before we can try to complete this type with
-      // another
-      if (!TypeSystemClang::DeclsAreEquivalent(const_cast<TagDecl *>(decl),
```
`TypeSystemClang::DeclsAreEquivalent` is now unused it seems. Lets remove it?

Michael137 reviewed (approved):
LGTM (just left some small comments)

clayborg commented on lldb/source/API/SBTarget.cpp:
```diff
@@ -1859,17 +1851,11 @@ lldb::SBTypeList SBTarget::FindTypes(const char *typename_cstr) {
+    TypeQuery query(typename_cstr);
+    TypeResults results;
+    images.FindTypes(nullptr, query, results);
+    for (const TypeSP &type_sp : results.GetTypeMap().Types())
```
I will change `TypeResults::InsertUnique(...)` to not add any invalid TypeSP objects.

clayborg commented on lldb/source/Commands/CommandObjectTarget.cpp:
```diff
@@ -1696,16 +1696,16 @@ static size_t LookupTypeInModule(Target *target,
+    SymbolContext sc;
+    if (module)
+      sc.module_sp = module->shared_from_this();
+    sc.SortTypeList(results.GetTypeMap(), type_list);
```
Will do

clayborg commented on lldb/include/lldb/Symbol/TypeSystem.h:
```diff
@@ -146,6 +133,9 @@ class TypeSystem : public PluginInterface,
   GetCompilerDeclContextForType(const CompilerType &type);
 
+  virtual std::vector<lldb_private::CompilerContext>
+  DeclContextGetCompilerContext(void *opaque_decl_ctx);
```
I would rather not put any limitations on TypeSystem plug-ins in case they can be more efficient if they can put off parsing information about a type until accessors are called on the `DeclContext` objects.

clayborg commented on lldb/include/lldb/Symbol/TypeSystem.h:
```diff
@@ -122,6 +106,9 @@ class TypeSystem : public PluginInterface,
                                                    size_t arg_idx);
 
+  virtual std::vector<lldb_private::CompilerContext>
+  DeclGetCompilerContext(void *opaque_decl);
```
I would rather not put any limitations on TypeSystem plug-ins in case they can be more efficient if they can put off parsing information about a type until accessors are called on the `Decl` objects.

Michael137 replied:
yea makes sense, looks like that's what all the other TypeSystem APIs do too

clayborg commented on lldb/source/Plugins/SymbolFile/PDB/SymbolFilePDB.cpp:
```diff
@@ -1446,23 +1446,23 @@ void SymbolFilePDB::AddSymbols(lldb_private::Symtab &symtab) {
+
+//   // There is an assumption 'name' is not a regex
+//   FindTypesByName(name.GetStringRef(), parent_decl_ctx, max_matches, types);
+// }
```
yes!

Michael137 commented:
BTW, lets prefix the commit with `[lldb]` before merging

clayborg committed:
Address all review comments:

Fixes include:
- fixing all header comments that had typos
- remove dead code from TypeSystemClang.cpp/.h
- remove commented out code in SymbolFilePDB.cpp
- add comment to CommandObjectTarget.cpp to clarify why type sorting is needed
- make sure we don't add any NULL TypeSP values in TypeResults::InsertUnique(...)
- remove debug statements from lldb/test/API/lang/cpp/unique-types4/TestUniqueTypes4.py

clayborg commented:
> BTW, lets prefix the commit with `[lldb]` before merging

Do we need to fix this in the commit message when I squash? Or do I just update the title of this PR?

clayborg commented:
> > > Sorry, it's been a while since I last reviewed this and I must admit I forgot about this in the mean time. The only thing I'd like to check before we can land this is to hear from @Michael137 who has been actively working on FindTypes performance recently whether this aligns or clashes with what he's been working on so we don't get a mid-air collision between patches.
> > 
> > 
> > Sounds good, I look forward to hearing if this fixes the issues as well. Before this fix I ran a case where I did an expression like "MyClass::iterator" and before this fix, with linux and no accelerator tables, we would convert DWARF into clang AST types for over 2 million types, and after this fix it went down to 11. This was when debugging LLDB + LLVM + clang with debug info, so there were TONS of "iterator" entries in the manual DWARF index.
> > On Apple the type accelerator tables have the hash of the fully qualified name which can help reduce the number of types we parse, but we don't have this on linux even with DWARF5 + .debug_names
> 
> To add more context to @adrian-prantl's comment, we've been working on reviving https://reviews.llvm.org/D101950, which changes the way we complete record types. The main blocker was performance, because as you mention, we often end up doing these expensive namespace searches, which gets much more noticeable with D101950. The problem we saw was mostly triggered by the second lookup that `-gsimple-template-names` does in `FindTypes`.
> 
> I applied your patch on top of D101950 and performance seems to be much better (by 2 orders of magnitude). It did break some tests locally for me, so I'm just skimming this patch to see what changed

All tests passed on my machine from with this, so you might try those tests without D101950 and see how things go? Happy to help if you need info on how to do things with the new lookups.

Michael137 commented:
> > > > Sorry, it's been a while since I last reviewed this and I must admit I forgot about this in the mean time. The only thing I'd like to check before we can land this is to hear from @Michael137 who has been actively working on FindTypes performance recently whether this aligns or clashes with what he's been working on so we don't get a mid-air collision between patches.
> > > 
> > > 
> > > Sounds good, I look forward to hearing if this fixes the issues as well. Before this fix I ran a case where I did an expression like "MyClass::iterator" and before this fix, with linux and no accelerator tables, we would convert DWARF into clang AST types for over 2 million types, and after this fix it went down to 11. This was when debugging LLDB + LLVM + clang with debug info, so there were TONS of "iterator" entries in the manual DWARF index.
> > > On Apple the type accelerator tables have the hash of the fully qualified name which can help reduce the number of types we parse, but we don't have this on linux even with DWARF5 + .debug_names
> > 
> > 
> > To add more context to @adrian-prantl's comment, we've been working on reviving https://reviews.llvm.org/D101950, which changes the way we complete record types. The main blocker was performance, because as you mention, we often end up doing these expensive namespace searches, which gets much more noticeable with D101950. The problem we saw was mostly triggered by the second lookup that `-gsimple-template-names` does in `FindTypes`.
> > I applied your patch on top of D101950 and performance seems to be much better (by 2 orders of magnitude). It did break some tests locally for me, so I'm just skimming this patch to see what changed
> 
> All tests passed on my machine from with this, so you might try those tests without D101950 and see how things go? Happy to help if you need info on how to do things with the new lookups.

All tests are passing for me now, it had to do with some badly resolved merge conflicts

> Do we need to fix this in the commit message when I squash? Or do I just update the title of this PR?

I think when you press the `Squash and merge` button it will give you the chance to correct the title of the commit

adrian-prantl reviewed (approved):
Given that Michael is okay with it, too, let's land this!

clayborg commented:
> Given that Michael is okay with it, too, let's land this!

I am testing on linux and windows to make sure everything is working and I will land this soon!

clayborg committed:
Fix a testcase so it passes on linux.

clayborg commented:
How did all of these commits get added to my PR???

clayborg commented:
I am going to abandon and resubmit in another PR...

clayborg commented:
Never mind, ran `git fetch upstream && git rebase upstream/main` and it fixed everything...

clayborg commented:
Buildbots are failing due to an unexpeted pass!

https://github.com/llvm/llvm-project/pull/75272

DavidSpickett commented:
I merged your fix and a few follow ups to get it all to compile. Now we have 2 failures on Windows:
https://lab.llvm.org/buildbot/#/builders/219/builds/7554/steps/6/logs/stdio

```
Failed Tests (2):
  lldb-api :: lang/cpp/unique-types4/TestUniqueTypes4.py
  lldb-unit :: SymbolFile/PDB/./SymbolFilePDBTests.exe/SymbolFilePDBTests/TestMaxMatches
```

Which I can look into but not today.

Is it expected that results would change here? Seems like a refactor but I don't have much context.

clayborg commented:
> I merged your fix and a few follow ups to get it all to compile. Now we have 2 failures on Windows: https://lab.llvm.org/buildbot/#/builders/219/builds/7554/steps/6/logs/stdio
> 
> ```
> Failed Tests (2):
>   lldb-api :: lang/cpp/unique-types4/TestUniqueTypes4.py
>   lldb-unit :: SymbolFile/PDB/./SymbolFilePDBTests.exe/SymbolFilePDBTests/TestMaxMatches
> ```
For TestUniqueTypes4.py, it is failing with a expression error, probably because a type lookup failed where it used to succeed.

TestMaxMatches will need to be debugged a bit. It used to find all of the types first, and then try to find a max of 1 through 10 types after that. Now with type queries, you get only one or all types (no one was using a max value for types other than 1 or all before). I did my best to try and code it up, but I have no ability to run the test to verify it works or debug it. Any help here would be appreciated. I do two type queries in the test where I grab all types and we expect more than one match, and then one where we try to find just one and then verify we got only one from the type map.


> 
> Which I can look into but not today.
> 
> Is it expected that results would change here? Seems like a refactor but I don't have much context.

Yeah, I don't have the ability yet to debug on windows. I am working on getting a VM based solution running now.

The new function `SymbolFilePDB::FindTypes(...)` that replaces the old one needs to be debugged. I took a guess at what would work, but not sure I got it right. Any help debugging this would be very much appreciated. This patch is a HUGE perf win, so it is worth making sure we can keep this in and making it work. If we need to expected fail the python test on window until we can look into it, I would vote to do this. And also fixing TextMaxMatches by commenting out this test for now until we can debug it would be a good option. With the supposition that we get this fixed within a week for both.

clayborg commented:
Debugging the SymbolFilePDBTests function would be the easiest way to reproduce and if the `TestMaxMatches` gets fixed, it might fix the other failing python test.

adrian-prantl commented:
@clayborg Do you by any chance happen to already have a version of this patch that applies cleanly to apple/llvm-project stable/20230725?
If not @Michael137 might try to backport it.

clayborg commented:
> @clayborg Do you by any chance happen to already have a version of this patch that applies cleanly to apple/llvm-project stable/20230725? If not @Michael137 might try to backport it.

I do not, only what is in top of tree. Totally worth backporting this though. I was able to most apply this patch cleanly from my original patch from early this year, so it shouldn't be too hard to backport.

DavidSpickett commented:
> If we need to expected fail the python test on window until we can look into it, I would vote to do this.

As it happens, I'm using the buildbot machine to reproduce this. So it's not doing any more builds for now. I'll put in workarounds if I don't find the issue by Friday.

clayborg commented:
> > If we need to expected fail the python test on window until we can look into it, I would vote to do this.
> 
> As it happens, I'm using the buildbot machine to reproduce this. So it's not doing any more builds for now. I'll put in workarounds if I don't find the issue by Friday.

Thank you. I tried installing Window on Parallels Desktop 19 on my mac but the VS 2019 doesn't run on arm64 macs, so I couldn't even build the latest upstream. 

I also was able to build on a windows SSH, but there are no command line debuggers for windows that I know of. I tried debugging by building LLDB on here and was able to debug, but not effectively enough to figure anything out.

clayborg commented:
So regarding:
```
lldb-unit :: SymbolFile/PDB/./SymbolFilePDBTests.exe/SymbolFilePDBTests/TestMaxMatches
```
This test was testing the wrong thing. From the looks of `SymbolFilePDBTests::TestMaxMatches(...)` I thought it was finding multiple matches when we didn't ask for a limit, and then it would iterate over the number of actual matches up to 10 and see if they found the exact number of matches. The problem is there is only one match for `ClassTypedef` (see `lldb/unittests/SymbolFile/PDB/Inputs/test-pdb-types.cpp` to verify). The old code would ask for all matches and only find 1, then it would iterate from 0 to < 1, so it would just call `FindTypes(...)` again and find that we have 1 of 1 matches. I figured that the test has multiple `ClassTypedef` entries in different scopes, but that isn't the case. So to fix this test I just need to fix `SymbolFilePDBTests::TestMaxMatches(...)` to verify there is 1 match in both cases. This test could be improved to actually test what it is supposed to. That will involve adding multiple `ClassTypedef` entries at different scopes (inside a namespace, in a class, and at the top level), but I will fix it for now. I also found an inefficient loop in `SymbolFilePDB::FindTypes(...)` that I will fix along with the fix.

clayborg commented:
@DavidSpickett I created a new PR to fix the buildbots:

https://github.com/llvm/llvm-project/pull/75566

This disables the TestUniqueTypes4.py on windows, and fixes the test in `lldb/unittests/SymbolFile/PDB/SymbolFilePDBTests.cpp` to test what it was testing before.

dwblaikie commented:
FWIW, we're seeing this breaking a pretty printer in google - previously a lookup was able to find `absl::cord_internal::RefcountAndFlags::Flags::kNumFlags` ( https://github.com/abseil/abseil-cpp/blob/8184f16e898fcb13b310b40430e13ba40679cf0e/absl/strings/internal/cord_internal.h#L202 ) (the name components are, in order: namespace, namespace, class, unscoped enumeration, enumerator)

The code we have (working via [gala](https://github.com/sivachandra/gala)) has been doing this query twice to workaround this bug: https://github.com/llvm/llvm-project/issues/53904 - but it looks like since this change, that workaround has even broken and the name is no longer found?

I don't have a fully isolated repro yet, but figured I'd start with raising the flag in case anyone recognizes this.

dwblaikie commented:
@clayborg any thoughts on this ^ ?

clayborg commented:
> absl::cord_internal::RefcountAndFlags::Flags::kNumFlags

I have a small repro case from the code pointer you gave. I will look into this and report back what I find. 

Are you guys using the expression parser in your pretty printer? As Michael pointed out, you can get the enum type as a `lldb::SBType` in and explore the enumerators via the API. Above Michael used a variable to get the type, but you can use the `lldb::SBType lldb::SBTarget::FindFirstType(const char *type)` API:
```
>>> t = lldb.target.FindFirstType("absl::cord_internal::RefcountAndFlags::Flags")
>>> t
enum Flags {
    kNumFlags,
    kImmortalFlag,
    kRefIncrement
}
>>> t.GetEnumMembers()
<lldb.SBTypeEnumMemberList; proxy of <Swig Object of type 'lldb::SBTypeEnumMemberList *' at 0x107d0fdb0> >
>>> t.GetEnumMembers()['kNumFlags']
unsigned int kNumFlags
>>> t.GetEnumMembers()['kNumFlags'].GetValueAsUnsigned()
1
```

That being said, it does seem like this issue is due to https://github.com/llvm/llvm-project/issues/53904. Getting the type from the target and getting the enum value directly will be much more efficient than using the expression parser, but we do need to fix the expression parser as I wasn't aware of this bug where running the expression multiple times would eventually allow it to work.

clayborg commented:
I am also not saying this patch hasn't made this issue happen more often, so I will check out my APIs to see how they are being called from the expression parser.

clayborg commented:
So I traced down why things used to work and why they don't work now. The issue is we are being too accurate with our type lookups now. For my reduced example I had this code:
```
struct A {
  struct B {
    static int f;
  };
};
```
What would happen when you type "p A::B::f" was this:
We would look for 'A' in the root namespace and we would find it, the record decl looks like:
```
CXXRecordDecl 0x105504e08 <<invalid sloc>> <invalid sloc> struct A definition
`-DefinitionData pass_in_registers empty aggregate standard_layout trivially_copyable pod trivial literal has_constexpr_non_copy_move_ctor can_const_default_init
  |-DefaultConstructor exists trivial constexpr needs_implicit defaulted_is_constexpr
  |-CopyConstructor simple trivial has_const_param needs_implicit implicit_has_const_param
  |-MoveConstructor exists simple trivial needs_implicit
  |-CopyAssignment simple trivial has_const_param needs_implicit implicit_has_const_param
  |-MoveAssignment exists simple trivial needs_implicit
  `-Destructor simple irrelevant trivial needs_implicit
```
We would look for 'B' in the root namespace, but since type lookups used to realize all basename matches (where 'B' would match 'A::B') and then we would throw away results that didn't match, but we did end up converting the A::B and it now exists in the 'A' CXXRecordDecl, but the compiler isn't looking for it there, so the expression fails with:
```
error: <user expression 0>:1:4: no member named 'B' in 'A'
    1 | A::B::f
      | ~~~^
```

When you run the expression the second time, the "struct A" now has a forward declaration for "A::B" in the CXXRecordDecl. So the second expression would look for 'A' in the root namespace and we would find it, the record decl 
 now looks like:
```
CXXRecordDecl 0x156b78a08 <<invalid sloc>> <invalid sloc> struct A definition
|-DefinitionData pass_in_registers empty aggregate standard_layout trivially_copyable pod trivial literal has_constexpr_non_copy_move_ctor can_const_default_init
| |-DefaultConstructor exists trivial constexpr needs_implicit defaulted_is_constexpr
| |-CopyConstructor simple trivial has_const_param needs_implicit implicit_has_const_param
| |-MoveConstructor exists simple trivial needs_implicit
| |-CopyAssignment simple trivial has_const_param needs_implicit implicit_has_const_param
| |-MoveAssignment exists simple trivial needs_implicit
| `-Destructor simple irrelevant trivial needs_implicit
`-CXXRecordDecl 0x156b78b58 <<invalid sloc>> <invalid sloc> <undeserialized declarations> struct B
```
Note the last line is our definition for 'A::B' that was not there before, but now since it is there, the compiler will find it.

With the current upstream we run and we find "A" and have the same CXXRecordDecl as the first expression in the old LLDB, but now we will ask for "B" in the root namespace and we won't find anything because we are smart enough to  _not_ create every type whose basename matches "B", so we won't create "B", so it won't get added to A's CXXRecordDecl that this causes the expression to never be able to succeed.

So I am surprised to see that the compiler is asking for "B" in the root namespace, and not asking us to find "B" in the "A" decl context. 

But the real issue might be that LLDB, which is acting like a precompiled header, and when asked us to find 'A', we return a forward declaration that is marked as "can be completed lazily via the precompiled header (external AST source) interface. And when we are asked to complete the 'A' type, it might expect us to add all contained types, like "struct A::B". We are not doing that right now, but maybe that is how the actual precompiled header support does things.
